src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,105 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "Utilities/TypeTraits/GetFundamentalType.hpp"
+#include "Utilities/TypeTraits/IsIterable.hpp"
+/**
+ * \ingroup DataStructuresGroup
+ *
+ * \brief A dynamically sized vector of DataVectors or ComplexDataVectors. If T
+ * is a floating point type, `DynamicBuffer<T>` resolves to std::vector<T>.
+ *
+ * \details This class is useful when one wants to create a `std::vector<T>`
+ * with a size that is unknown at compile time. It allocates all DataVectors in
+ * a single memory chunk rather than allocating each individually. If the size
+ * of the vector is known at compile time, a `TempBuffer` object should be used
+ * instead.
+ */
+
+template <typename T>
+class DynamicBuffer {
+ public:
+  static constexpr bool is_data_vector_type =
+      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T>, T>;
+
+  DynamicBuffer() = default;
+
+  /*
+   * Constructs a `DynamicBuffer`. The `outer_size` is the size of the outer
+   * `std::vector`, the `inner_size` is the size of the objects of type T stored
+   * inside the DataVector. The `inner_size` has to be size 1 if T is a
+   * fundamental type.
+   */
+  DynamicBuffer(const size_t outer_size, const size_t inner_size) noexcept
+      : inner_size_(inner_size), data_vectors_(outer_size) {
+    if constexpr (is_data_vector_type) {
+      buffer_.resize(outer_size * inner_size_);
+      set_references();
+    } else {
+      if (inner_size != 1) {
+        ERROR(
+            "DynamicBuffer must have inner_size == 1 when T is a fundamental "
+            "type but has inner size "
+            << inner_size_);
+      }
+    }
+  }
+
+  ~DynamicBuffer() = default;
+  DynamicBuffer(const DynamicBuffer& other) noexcept
+      : inner_size_(other.inner_size_) {
+    if constexpr (is_data_vector_type) {
+      data_vectors_.resize(other.size());
+      buffer_ = other.buffer_;
+      set_references();
+    } else {
+      data_vectors_ = other.data_vectors_;
+    }
+  }
+  DynamicBuffer(DynamicBuffer&& other) noexcept
+      : inner_size_(other.inner_size_) {
+    if constexpr (is_data_vector_type) {
+      data_vectors_.resize(other.size());
+      buffer_ = std::move(other.buffer_);
+      set_references();
+    } else {
+      data_vectors_ = std::move(other.data_vectors_);
+    }
+  }
+
+  DynamicBuffer& operator=(DynamicBuffer other) noexcept {
```

Comment:
This is the heart of the copy-and-swap idiom. By taking the parameter by value it acts as both the copy assignment operator and the move assignment operator. When called with an lvalue, the copy is made in the parameter list using the copy constructor, when called with an rvalue it will be move constructed. `std::swap` then uses the move constructor to swap the two objects around.

 Even nicer, it makes self-assignment checks redundant and (not that we care) guarantees strong exception safety.

Reply:
How does this make self-assignment checks redundant? You perform an entire copy when you self-assign, or how does the copy constructor call magically get elided? I'd argue it also doesn't guarantee exception safety because the copy constructor doesn't guarantee that. That is, if I do `a = b;` this can throw because the copy constructor can throw, so nothing gained.

My general view on this is it's trying to be "too clever" where the simpler, standard code is easier to understand and has fewer nuances. The general advice is to _not_ do this when there's heap-allocated memory and write an actual copy and move assignment operators, so I'd prefer we do that.

Reply:
@wthrowe do you have any insight here?

Reply:
The posted code doesn't work.  If you try to assign the class you'll overflow the stack because swap performs assignments.

Reply:
@nilsdeppe yes it will still make the copy. It is my understanding that self assignment checks are there to prevent proper ill-defined programs (e.g. when deleting the resource and then assigning to it), not to optimize away the copy. Self-assignment should be extremely rare and performing the check every call introduces its own cost.
Of course it can also still throw, it guarantees strong exception safety because all resources will be left intact. It does not guarantee no-throw as it necessarily allocates. This doesn't matter for spectre anyway it was just a side note.

@wthrowe you are right. The unit test did not catch that because I had naively assumed that `DynamicBuffer<T> dynamic_buffer_assigned = *dynamic_buffer;` would call the copy assignment operator but it just calls the copy constructor. For this to work properly I should not have removed the custom swap method or introduced a defaulted move assignment operator. 

I have changed it to a straight-forward implementation now. However, I do think that in general using very common design patterns such as the copy-and-swap idiom should not be discouraged. They are recommended pretty much anywhere you look when it comes to managing resources exactly because all nuances have been understood very well and because directly implementing all constructors has its own pitfalls.

Reply:
- Yes, in general, we do have some places where self-assignment is done frequently, e.g. with the DataBox. In generic code it's not always all that obvious how frequently self-assignment will be done :Shrug: Of course, an if-statement unless evaluated in a tight loop is not exactly a huge cost in general either. Much lower than the dozens to hundreds of memory allocations this class saves :)
- Okay...
- It seems like in this case the idiom caused more problems than just implementing copy assignment, so it's not a very convincing case that this idiom is something that should be adopted :) As I mentioned before, I don't remember this idiom being the recommended one in any of the books I've read, but maybe I'm just not remembering. It seems like the place where this idiom provides a benefit is when a custom copy and move are required, but if only a custom copy is needed it seems a lot less useful, especially if a custom swap function needs to be written.